proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,637 @@
+of associated token account. A associated token account is a `PDA` derived from
+associated token program, where the owner is token program. But the program
+internally saves who has the authority over the account. Only token program can
+change the data of a token account. For operations like withdrawing tokens, the
+authority of the associated token account has to be the signer for the
+transaction containing transfer instruction. To receive tokens from another
+account, the authority's signature is not required.
+
+`Other Fees`: Fees other than PRAW fees `base fees + prioritization fees`.
+
+## Detailed Design
+
+PRAW fees enable dapp developer to decide who to rebate and who to penalize
+through runtime features. In particular, programs needs to be able to penalize,
+even if the application is not CPI'd to.
+
+The checks on the PRAW fees will be taken care of by Solana runtime. Total PRAW
+fees committed will be included in the transaction through a special parameter
+called `SetProgramRebatableAccountWriteFees` so that it is easier to calculate
+the total amount of fees that will be paid by the transaction. This parameter
+will be part of a new `TransactionHeaderProgram` where all the transaction
+parameters will be set through a TVL. In case of nonpayment or partial payment,
+the program is never executed. This instruction cannot be CPI'ed into.
+
+Addition of new syscalls to get, set and rebate PRAW fees during program
+runtime. Application developers will be able to rebate these fees through the
+new syscall, but only if the transaction succeeds, this way transactions
+write-locking the respective accounts will require full fee payment before
+starting the execution of the transaction. Rebates will be issued to the payer
+after the transaction is executed. During transaction execution, the rebates
+will not be reflected in the payer's balance.
+
+Once PRAW fees are paid on an account by the payer they are valid for the whole
+transaction, even if the same account is write-locked in multiple instructions.
+
+### Payment
+
+`SetProgramRebatableAccountWriteFees` will be part of the updated compute budget
+program. `SetProgramRebatableAccountWriteFees` will take a u64 parameter to set
+lamports per requested CU as the PRAW fees. Instead of having separate
+instructions for each compute budget parameter and PRAW fee we will have a
+single instruction that will set all the compute budget and PRAW fee parameters.
+Compute budget parameters could be considered as transaction header that sets
+some parameters for the transaction. We create a new solana native program
+`TransactionHeaderProgram` that decodes the instruction data using a type length
+value structure (`TLV`) format and computes the required
+`TransactionHeaderParameter`s for the transaction. The `TLV` structure is well
+defined for accounts in solana program library repository we will extend the
+definition for instruction data. As we have multiple transaction header
```

Comment:
Anyways we need to add a way to set PRAW fees. It will be harder to set in compute budget, would be better if it is a part of compute budget instruction. That is why I made it a part of this proposal.